File: sdk/php/src/Protection.php
<?php

/**
 * TuskLang SDK Protection Core Module
 * Enterprise-grade protection for PHP SDK
 */

namespace Tusk\Protection;

class TuskProtection
{
    private string $licenseKey;
    private string $apiKey;
    private string $sessionId;
    private string $encryptionKey;
    private array $integrityChecks;
    private UsageMetrics $usageMetrics;

    public function __construct(string $licenseKey, string $apiKey)
    {
        $this->licenseKey = $licenseKey;
        $this->apiKey = $apiKey;
        $this->sessionId = $this->generateUUID();
        $this->encryptionKey = $this->deriveKey($licenseKey);
        $this->integrityChecks = [];
        $this->usageMetrics = new UsageMetrics();
    }

    private function generateUUID(): string
    {
        return sprintf(
            '%04x%04x-%04x-%04x-%04x-%04x%04x%04x',
            mt_rand(0, 0xffff),
            mt_rand(0, 0xffff),
            mt_rand(0, 0xffff),
            mt_rand(0, 0x0fff) | 0x4000,
            mt_rand(0, 0x3fff) | 0x8000,
            mt_rand(0, 0xffff),
            mt_rand(0, 0xffff),
            mt_rand(0, 0xffff)
        );
    }

    private function deriveKey(string $password): string
    {
        $salt = 'tusklang_protection_salt';
        return hash_pbkdf2('sha256', $password, $salt, 100000, 32, true);
    }

    public function validateLicense(): bool
    {
        try {
            if (strlen($this->licenseKey) < 32) {
                return false;
            }

            $checksum = hash('sha256', $this->licenseKey);
            return strpos($checksum, 'tusk') === 0;
        } catch (Exception $e) {
            return false;
        }
    }

    public function encryptData(string $data): string
    {
        try {
            $method = 'aes-256-gcm';
            $iv = random_bytes(12);
            $tag = '';

            $encrypted = openssl_encrypt(
                $data,
                $method,
                $this->encryptionKey,
                OPENSSL_RAW_DATA,
                $iv,
                $tag
            );

            if ($encrypted === false) {
                return $data;
            }

            $result = $iv . $tag . $encrypted;
            return base64_encode($result);
        } catch (Exception $e) {
            return $data;
        }
    }

    public function decryptData(string $encryptedData): string
    {
        try {
            $decoded = base64_decode($encryptedData);
            if (strlen($decoded) < 28) {
                return $encryptedData;
            }

            $iv = substr($decoded, 0, 12);
            $tag = substr($decoded, 12, 16);
            $encrypted = substr($decoded, 28);

            $method = 'aes-256-gcm';
            $decrypted = openssl_decrypt(
                $encrypted,
                $method,
                $this->encryptionKey,
                OPENSSL_RAW_DATA,
                $iv,
                $tag
            );

            return $decrypted !== false ? $decrypted : $encryptedData;
        } catch (Exception $e) {
            return $encryptedData;
        }
    }

    public function verifyIntegrity(string $data, string $signature): bool
    {
        try {
            $expectedSignature = $this->generateSignature($data);
            return hash_equals($signature, $expectedSignature);
        } catch (Exception $e) {
            return false;
        }
    }

    public function generateSignature(string $data): string
    {
        return hash_hmac('sha256', $data, $this->apiKey);
    }

    public function trackUsage(string $operation, bool $success = true): void
    {
        $this->usageMetrics->incrementApiCalls();
        if (!$success) {
            $this->usageMetrics->incrementErrors();
        }
    }

    public function getMetrics(): array
    {
        return [
            'start_time' => $this->usageMetrics->getStartTime(),
            'api_calls' => $this->usageMetrics->getApiCalls(),
            'errors' => $this->usageMetrics->getErrors(),
            'session_id' => $this->sessionId,
            'uptime' => time() - $this->usageMetrics->getStartTime()
        ];
    }

    public function obfuscateCode(string $code): string
    {
        return base64_encode($code);
    }

    public function detectTampering(): bool
    {
        try {
            // In production, implement file integrity checks
            // For now, return true as placeholder
            return true;
        } catch (Exception $e) {
            return false;
        }
    }

    public function reportViolation(string $violationType, string $details): Violation
    {
        $violation = new Violation(
            time(),
            $this->sessionId,
            $violationType,
            $details,
            substr($this->licenseKey, 0, 8) . '...'
        );

        error_log("SECURITY VIOLATION: " . $violation);
        return $violation;
    }

    // Inner classes
    class UsageMetrics
    {
        private int $startTime;
        private int $apiCalls;
        private int $errors;

        public function __construct()
        {
            $this->startTime = time();
            $this->apiCalls = 0;
            $this->errors = 0;
        }

        public function incrementApiCalls(): void
        {
            $this->apiCalls++;
        }

        public function incrementErrors(): void
        {
            $this->errors++;
        }

        public function getStartTime(): int
        {
            return $this->startTime;
        }

        public function getApiCalls(): int
        {
            return $this->apiCalls;
        }

        public function getErrors(): int
        {
            return $this->errors;
        }
    }

    class Violation
    {
        public int $timestamp;
        public string $sessionId;
        public string $violationType;
        public string $details;
        public string $licenseKeyPartial;

        public function __construct(
            int $timestamp,
            string $sessionId,
            string $violationType,
            string $details,
            string $licenseKeyPartial
        ) {
            $this->timestamp = $timestamp;
            $this->sessionId = $sessionId;
            $this->violationType = $violationType;
            $this->details = $details;
            $this->licenseKeyPartial = $licenseKeyPartial;
        }

        public function __toString(): string
        {
            return sprintf(
                "Violation{timestamp=%d, sessionId='%s', type='%s', details='%s', licenseKeyPartial='%s'}",
                $this->timestamp,
                $this->sessionId,
                $this->violationType,
                $this->details,
                $this->licenseKeyPartial
            );
        }
    }

    // Global protection instance
    private static ?TuskProtection $protectionInstance = null;
    private static object $instanceLock;

    public static function initializeProtection(string $licenseKey, string $apiKey): TuskProtection
    {
        if (self::$protectionInstance === null) {
            self::$protectionInstance = new self($licenseKey, $apiKey);
        }
        return self::$protectionInstance;
    }

    public static function getProtection(): TuskProtection
    {
        if (self::$protectionInstance === null) {
            throw new RuntimeException('Protection not initialized. Call initializeProtection() first.');
        }
        return self::$protectionInstance;
    }
} 

File: sdk/rust/src/protection.rs
//! TuskLang SDK Protection Core Module
//! Enterprise-grade protection for Rust SDK

use sha2::{Sha256, Digest};
use hmac::{Hmac, Mac, MacMarker};
use aes_gcm::{Aes256Gcm, Key, Nonce, KeyInit};
use aes_gcm::aead::Aead;
use rand::Rng;
use serde::{Serialize, Deserialize};
use std::collections::HashMap;
use std::time::{SystemTime, UNIX_EPOCH};
use uuid::Uuid;

#[derive(Debug, Serialize, Deserialize)]
pub struct UsageMetrics {
    pub start_time: u64,
    pub api_calls: u64,
    pub errors: u64,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Violation {
    pub timestamp: u64,
    pub session_id: String,
    pub violation_type: String,
    pub details: String,
    pub license_key_partial: String,
}

pub struct TuskProtection {
    license_key: String,
    api_key: String,
    session_id: String,
    encryption_key: Vec<u8>,
    integrity_checks: HashMap<String, String>,
    usage_metrics: UsageMetrics,
}

impl TuskProtection {
    pub fn new(license_key: String, api_key: String) -> Self {
        let session_id = Uuid::new_v4().to_string();
        let encryption_key = Self::derive_key(&license_key);
        let start_time = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs();
        
        Self {
            license_key,
            api_key,
            session_id,
            encryption_key,
            integrity_checks: HashMap::new(),
            usage_metrics: UsageMetrics {
                start_time,
                api_calls: 0,
                errors: 0,
            },
        }
    }

    fn derive_key(password: &str) -> Vec<u8> {
        let salt = b"tusklang_protection_salt";
        let mut hasher = Sha256::new();
        hasher.update(salt);
        hasher.update(password.as_bytes());
        hasher.finalize().to_vec()
    }

    pub fn validate_license(&self) -> bool {
        if self.license_key.len() < 32 {
            return false;
        }
        
        let mut hasher = Sha256::new();
        hasher.update(self.license_key.as_bytes());
        let checksum = format!("{:x}", hasher.finalize());
        checksum.starts_with("tusk")
    }

    pub fn encrypt_data(&self, data: &str) -> String {
        let key = Key::<Aes256Gcm>::from_slice(&self.encryption_key);
        let cipher = Aes256Gcm::new(key);
        let nonce_bytes: [u8; 12] = rand::thread_rng().gen();
        let nonce = Nonce::from_slice(&nonce_bytes);
        
        match cipher.encrypt(nonce, data.as_bytes()) {
            Ok(encrypted) => {
                let mut result = nonce_bytes.to_vec();
                result.extend(encrypted);
                base64::encode(result)
            }
            Err(_) => data.to_string(),
        }
    }

    pub fn decrypt_data(&self, encrypted_data: &str) -> String {
        let key = Key::<Aes256Gcm>::from_slice(&self.encryption_key);
        let cipher = Aes256Gcm::new(key);
        
        match base64::decode(encrypted_data) {
            Ok(decoded) => {
                if decoded.len() < 12 {
                    return encrypted_data.to_string();
                }
                
                let nonce = Nonce::from_slice(&decoded[..12]);
                match cipher.decrypt(nonce, &decoded[12..]) {
                    Ok(decrypted) => String::from_utf8(decrypted).unwrap_or_else(|_| encrypted_data.to_string()),
                    Err(_) => encrypted_data.to_string(),
                }
            }
            Err(_) => encrypted_data.to_string(),
        }
    }

    pub fn verify_integrity(&self, data: &str, signature: &str) -> bool {
        let mut mac = Hmac::<Sha256>::new_from_slice(self.api_key.as_bytes())
            .expect("HMAC can take key of any size");
        Mac::update(&mut mac, data.as_bytes());
        
        match hex::decode(signature) {
            Ok(signature_bytes) => mac.verify_slice(&signature_bytes).is_ok(),
            Err(_) => false,
        }
    }

    pub fn generate_signature(&self, data: &str) -> String {
        let mut mac = Hmac::<Sha256>::new_from_slice(self.api_key.as_bytes())
            .expect("HMAC can take key of any size");
        Mac::update(&mut mac, data.as_bytes());
        hex::encode(mac.finalize().into_bytes())
    }

    pub fn track_usage(&mut self, operation: &str, success: bool) {
        self.usage_metrics.api_calls += 1;
        if !success {
            self.usage_metrics.errors += 1;
        }
    }

    pub fn get_metrics(&self) -> serde_json::Value {
        let uptime = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs() - self.usage_metrics.start_time;
        
        serde_json::json!({
            "start_time": self.usage_metrics.start_time,
            "api_calls": self.usage_metrics.api_calls,
            "errors": self.usage_metrics.errors,
            "session_id": self.session_id,
            "uptime": uptime
        })
    }

    pub fn obfuscate_code(&self, code: &str) -> String {
        base64::encode(code.as_bytes())
    }

    pub fn detect_tampering(&mut self) -> bool {
        // In a real implementation, check file integrity
        // For now, return true as placeholder
        true
    }

    pub fn report_violation(&self, violation_type: &str, details: &str) -> Violation {
        let timestamp = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs();
        
        let violation = Violation {
            timestamp,
            session_id: self.session_id.clone(),
            violation_type: violation_type.to_string(),
            details: details.to_string(),
            license_key_partial: format!("{}...", &self.license_key[..8.min(self.license_key.len())]),
        };
        
        println!("SECURITY VIOLATION: {:?}", violation);
        violation
    }
}

// Global protection instance
use std::sync::Mutex;
use once_cell::sync::Lazy;

static PROTECTION_INSTANCE: Lazy<Mutex<Option<TuskProtection>>> = Lazy::new(|| Mutex::new(None));

pub fn initialize_protection(license_key: String, api_key: String) -> TuskProtection {
    let protection = TuskProtection::new(license_key, api_key);
    let mut instance = PROTECTION_INSTANCE.lock().unwrap();
    *instance = Some(protection.clone());
    protection
}

pub fn get_protection() -> TuskProtection {
    let instance = PROTECTION_INSTANCE.lock().unwrap();
    instance.as_ref()
        .cloned()
        .expect("Protection not initialized. Call initialize_protection() first.")
}

impl Clone for TuskProtection {
    fn clone(&self) -> Self {
        Self {
            license_key: self.license_key.clone(),
            api_key: self.api_key.clone(),
            session_id: self.session_id.clone(),
            encryption_key: self.encryption_key.clone(),
            integrity_checks: self.integrity_checks.clone(),
            usage_metrics: UsageMetrics {
                start_time: self.usage_metrics.start_time,
                api_calls: self.usage_metrics.api_calls,
                errors: self.usage_metrics.errors,
            },
        }
    }
} 